File: src/KI_36/Proshkin/Lab2/Rifle.java
package KI_36.Proshkin.Lab2;

import java.io.IOException;

/**
 * Class representing a rifle with shot tracking, cooling, and cleaning logic.
 */
public class Rifle {
    private final String type;
    private final String manufacturer;
    private Bullet bullet;
    private final Magazine magazine;
    private int shotsAmount; // Field to track number of shots
    private boolean cooled;  // Field to track if rifle is cooled down
    private boolean cleaned; // Field to track if rifle is cleaned
    private final Logger logger;

    // Parameterized constructor
    public Rifle(String type, String manufacturer, Bullet bullet, Magazine magazine) throws IOException {
        this.type = type;
        this.manufacturer = manufacturer;
        this.bullet = bullet;
        this.magazine = magazine;
        this.shotsAmount = 0;  // Initialize shots amount to 0
        this.cooled = true;    // Initially cooled
        this.cleaned = true;   // Initially cleaned
        this.logger = new Logger();
        logger.write("Rifle created: " + this);
    }

    // 1. Method to shoot
    public void shoot(int shots) throws IOException {
        if (shotsAmount >= 50 && !cleaned) {
            logger.write("Cannot shoot: Rifle needs cleaning!");
            System.out.println("You need to clean the rifle before shooting again!");
        } else if (shotsAmount >= 20 && !cooled) {
            logger.write("Cannot shoot: Rifle is too hot and needs cooling!");
            System.out.println("Your rifle is too hot and needs to cool down before shooting again!");
        } else if (bullet.getQuantity() > 0) {
                bullet.use();
                shotsAmount++;  // Increment the number of shots fired
                cleaned = false;  // Mark as not cleaned after firing
                cooled = false;
                logger.write("Shot fired. Shots amount: " + shotsAmount + ". Remaining bullets: " + bullet.getQuantity());
                System.out.println("Shot fired! Shots amount: " + shotsAmount);

        } else {
            logger.write("No bullets left.");
            System.out.println("No bullets left.");
        }
    }

    // 2. Method to reload the rifle
    public void reload(int quantity) throws IOException {
        int checkAmount = bullet.getQuantity() + quantity;
        if(checkAmount <= magazine.getMaxAmountOfBullets())
        {
            bullet.add(quantity);
            logger.write("Reloaded. Total bullets: " + bullet.getQuantity());
        }
        else
        {
            logger.write("Too much bullets for this magazine.");
        }
    }

    // 3. Method to check ammunition
    public void checkAmmo() throws IOException {
        logger.write("Checked ammo: " + bullet.getQuantity());
    }

    // 4. Method to change bullets (e.g., change caliber)
    public void changeBullet(Bullet newBullet) throws IOException {
        this.bullet = newBullet;
        logger.write("Bullet changed to: " + newBullet.getCaliber());
    }

    // 5. Method to clean the rifle
    public void clean() throws IOException {
        logger.write("Rifle cleaned.");
        cleaned = true;   // Mark as cleaned
        System.out.println("Rifle cleaned. Shots amount reset.");
    }

    // 6. Method to get information about the rifle
    public void getInfo() throws IOException {
        String info = this.toString() +
                "\nShots fired: " + shotsAmount +
                "\nCooled down: " + cooled +
                "\nCleaned: " + cleaned;logger.write("Info retrieved: " + info);
        System.out.println(info);
    }

    // 7. Method to check if the rifle is functional
    public void isFunctional() throws IOException {
        boolean functional = bullet.getQuantity() > 0 && cooled && cleaned;
        System.out.println("Functional : " + functional);
        logger.write("Functional check: " + functional);
    }

    // 8. Method to change the firing mode (e.g., single/automatic)
    public void changeFireMode(String mode) throws IOException {
        logger.write("For now only Single mode available");
    }

    // 9. Method to cool down the rifle after use
    public void coolDown() throws IOException {
        logger.write("Rifle cooled down.");
        cooled = true;  // Mark as cooled
        System.out.println("Rifle cooled down.");
    }

    // Method to close the logger and clean up resources
    public void close() throws IOException {
        logger.write("Work with rifle finished.");
        logger.close();
    }

    //Method to collect basic info
    public String toString() {
        return "Rifle{" +
                "type='" + type + "' " +
        ", manufacturer='" + manufacturer + "'" +
        ", bullet=" + bullet +
                '}';
    }
}

File: Rifle.py
import Bullet
import Magazine

class Rifle:
    def __init__(self, type, manufacturer, bullet):
        self.type = type
        self.manufacturer = manufacturer
        self.magazine = Magazine.Magazine(40)
        self.bullet = bullet
        self.shots_amount = 0  # Initialize shots amount to 0
        self.cooled = True     # Initially cooled
        self.cleaned = True    # Initially cleaned

    def shoot(self):
        if self.shots_amount >= 50 and not self.cleaned:
            print("Cannot shoot: Rifle needs cleaning!")
            print("You need to clean the rifle before shooting again!")
        elif self.shots_amount >= 20 and not self.cooled:
            print("Cannot shoot: Rifle is too hot and needs cooling!")
            print("Your rifle is too hot and needs to cool down before shooting again!")
        elif self.bullet.get_quantity() > 0:
                self.bullet.use()
                self.shots_amount += 1
                self.cleaned = False
                print(f"Shot fired! Shots amount: {self.shots_amount}")

    def reload(self, quantity):
        """
        Reloads the rifle with a specified quantity of bullets.

        Args:
            quantity (int): The number of bullets to reload.

        Raises:
            ValueError: If the total bullets exceed the magazine's capacity.
        """
        if self.bullet.quantity + quantity <= self.magazine.get_max_amount_of_bullets():
            self.bullet.add(quantity)
            print(f"Reloaded. Total bullets: {self.bullet.get_quantity()}")
        else:
            print("Too many bullets for this magazine.")

    def check_ammo(self):
        """
        Logs the current quantity of bullets.
        """
        print(f"Checked ammo: {self.bullet.get_quantity()}")

    def change_bullet(self, new_bullet):
        """
        Changes the type of bullet used by the rifle.

        Args:
            new_bullet (Bullet): The new bullet to be used.
        """
        self.bullet = new_bullet
        print(f"Bullet changed to: {new_bullet.get_caliber()}")

    def clean(self):
        """
        Cleans the rifle, resetting its cleaned status.
        """
        self.cleaned = True        
        print("Rifle cleaned. Shots amount reset.")

    def get_info(self):
        """
        Prints and logs information about the rifle.
        """
        info = (
            f"{self}\n"
            f"Shots fired: {self.shots_amount}\n"
            f"Cooled down: {self.cooled}\n"
            f"Cleaned: {self.cleaned}"
        )
        print(info)

    def is_functional(self):
        """
        Checks if the rifle is functional.

        Returns:
            bool: True if the rifle is functional, False otherwise.
        """
        functional = self.bullet.get_quantity() < 20 and self.cooled and self.cleaned
        print(f"Functional: {functional}")
        return functional

    def change_fire_mode(self, mode):
        """
        Changes the firing mode of the rifle.

        Args:
            mode (str): The new firing mode (e.g., "Single", "Automatic").
        """
        valid_modes = ["Single", "Automatic", "Burst", "Full load"]
        if mode in valid_modes:
            self.fire_mode = mode
            print(f"Fire mode changed to: {mode}")
        else:
            print(f"Invalid fire mode: {mode}")

    def cool_down(self):
        """
        Cools down the rifle after use.
        """
        self.cooled = True
        print("Rifle cooled down.")

    def __str__(self):
        """
        Returns a string representation of the Rifle instance.

        Returns:
            str: A string representation of the rifle.
        """
        return f"Rifle(type='{self.type}', manufacturer='{self.manufacturer}', bullet={self.bullet})"
    
    def close(self):
        print(f"Rifle(type='{self.type}', manufacturer='{self.manufacturer}', bullet={self.bullet} Shots amount: {self.shots_amount}. Remaining bullets: {self.bullet.get_quantity()}")